### PR TITLE
Login: Enhance error states (JavaScript not enabled, browser support lacking)

### PR DESCRIPTION
**Pull request by garrett**
Follow-up to #14870 (and currently includes its commits). To be rebased after #14870 lands.

### CODE REVIEW

**Comment by garrett** [comment]
TODO:

- [ ] Fix up browser support messaging

**Comment by garrett** [comment]
I would've had to add hacks on hacks on hacks to hackily fix the error messages. It was a tower of cards.

Instead, I kept ripping out more and more code to make the CSS more approachable for both not-me and also me. :wink: Whoops! Feature creep. But it needed to be done for a few years now.

Bonus points: Background image is much closer to PF4 now and even smaller in filesize (as I optimized it a bunch), and CSS is much simpler and a smaller file too. The page should be even a bit quicker to load now too.

Obviously, we'll have to see what the tests say, and it will need some help with checking all the different modes.

I'm keeping this as a draft right now, as I want to go over it a bit more next week and still probably have a few things to fix. (See TODO above.)

**Comment by mvollmer** [comment]
When "Other options" is open, the text is too close to the triangle:

image

**Comment by mvollmer** [comment]
The alignment of the banner icon has changed from this

image

to this

image

Neither looks obviously correct to me... I would align it with the first line.

**Comment by garrett** [comment]
> When "Other options" is open, the text is too close to the triangle:

I didn't see that yesterday,  Perhaps I broke it in a later commit or when rebasing. I'll check again.

> Neither looks obviously correct to me... I would align it with the first line.

Yep, as would I. I didn't see the alert. Is there a good way to trigger it?

**Comment by mvollmer** [comment]
> Yep, as would I. I didn't see the alert. Is there a good way to trigger it?

I just did this:

```
--- a/src/ws/login.js
+++ b/src/ws/login.js
@@ -17,6 +17,7 @@ if (window.NodeList && !NodeList.prototype.forEach)
 
     var url_root;
     var environment = window.environment || { };
+    environment.banner = "First line\nSecond line\nThird line";
     var oauth = environment.OAuth || null;
     if (oauth) {
         if (!oauth.TokenParam)
```

File: src/ws/login.html [hunk 1]
```diff
@@ -96,7 +93,11 @@ <h4 id="login-info-message" class="pf-c-alert__title"></h4>
           <a id="login-again" translate hidden>Try again</a>
         </div>
 
-        <div class="unsupported-browser" id="unsupported-browser" hidden>
+        <div class="noscript" translate>
+          Please enable JavaScript to use the Web Console.
```
Comment (mvollmer) [inline src/ws/login.html:97]: What about styling this like an error message, similar to the message about missing features?

If JSON is missing, we get this:

image

When JS is disabled, it looks much plainer:

image

Also, isn't it confusing to tell people to download a modern browser when they probably already have one, just with JavaScript disabled?

**Review by mvollmer — CHANGES_REQUESTED** [review]
LGTM, just the two layout issues mentioned in the comments.

File: src/ws/login.html (hunk 1 above)
Comment (garrett) [reply src/ws/login.html:97]: > When JS is disabled, it looks much plainer

Yeah, I guess I should style that too. Good idea.

> Also, isn't it confusing to tell people to download a modern browser when they probably already have one, just with JavaScript disabled?

It depends on why they don't have JavaScript.

I thought if someone is using some weird browser or a browser that cannot handle the JS, suggesting other browsers makes sense. But perhaps it doesn't? I guess in most cases, it's due to an add-on blocking JS or sometimes even a setting in the browser to turn it off.

**Comment by garrett** [comment]
TODO:
- [x] Fix arrow placement (and put inside the link)
- [ ] Make missing JavaScript notice look like an error
- [ ] Hide browser text when there's no JavaScript (as it's most likely a JS blocker, not an old/broken browser)

**Comment by garrett** [comment]
## No JavaScript, desktop
Screenshot_2020-11-18 Loading
## Missing support, desktop
Screenshot_2020-11-18 rain

## No JavaScript, mobile-ish
Screenshot_2020-11-18 Loading (1)
## Missing support, mobile-ish
Screenshot_2020-11-18 rain(1)

**Comment by garrett** [comment]
## Alerts, desktop
Screenshot_2020-11-18 rain(2)

## Alerts, mobile-ish
Screenshot_2020-11-18 rain(3)

I, um, need to
- [ ] fix that top alert on mobile sizes.

**Comment by garrett** [comment]
## Alerts, desktop
Screenshot_2020-11-18 rain(2)

## Alerts, mobile-ish
Screenshot_2020-11-18 rain(3)

I, um, need to
- [ ] fix that top alert on mobile sizes.

**Comment by garrett** [comment]
## Alerts, mobile, fixed
Screenshot_2020-11-18 rain(4)

## Alerts, desktop, still works :wink: 
Screenshot_2020-11-18 rain(5)

**Comment by garrett** [comment]
I dropped the 10+ from Windows for Edge, as the Chromium edge supports Windows 7+ (even if Microsoft doesn't support Windows 7 and 8.0; they still support Windows 8.1 in extended support, and Edgeium supports that).

**Comment by mvollmer** [comment]
`test/verify/check-login TestLogin.testBasic` needs fixing now. I think just `"#banner pre"` -> `"#banner-message"` should be enough.

**Comment by mvollmer** [comment]
> `test/verify/check-login TestLogin.testBasic` needs fixing now. I think just `"#banner pre"` -> `"#banner-message"` should be enough.

Yes, this is enough:

```
--- a/test/verify/check-login
+++ b/test/verify/check-login
@@ -61,7 +61,7 @@ account    required     pam_succeed_if.so user ingroup %s""" % m.get_admin_group
         b.reload()
         b.wait_visible("#login-user-input")
         b.wait_visible("#banner")
-        self.assertEqual(b.text("#banner pre"), m.execute("cat /etc/issue").rstrip())
+        self.assertEqual(b.text("#banner-message"), m.execute("cat /etc/issue").rstrip())
 
         # Test non existent file
         m.execute("printf '[Session]\nBanner = /etc/non-existing-file\n' > /etc/cockpit/cockpit.conf")
```

**Review by mvollmer — APPROVED** [review]

**Comment by garrett** [comment]
Force-pushed a with the suggested test fix (as part of the alert commit)

**Comment by garrett** [comment]
Not sure why a few tests are still failing, but most pass. :shrug:

**Review by mvollmer — APPROVED** [review]

**Pull request merged**